src/kernel/retype_tbl.c

```diff
@@ -95,125 +227,235 @@ mod_mem_type(void *pa, const mem_type_t type)
+			if (walk[i].inc_cnt != 0) cos_faa((int*)walk[i].p_glb, -1);
+		}
+	}
+
+	return ret;
 }
 
 int
-retypetbl_retype2user(void *pa)
+retypetbl_retype2user(void *pa, u32_t order)
 {
-	return mod_mem_type(pa, RETYPETBL_USER);
+	return mod_mem_type(pa, order, RETYPETBL_USER);
 }
 
 int
-retypetbl_retype2kern(void *pa)
+retypetbl_retype2kern(void *pa, u32_t order)
 {
-	return mod_mem_type(pa, RETYPETBL_KERN);
+	return mod_mem_type(pa, order, RETYPETBL_KERN);
 }
 
 /* implemented in pgtbl.c */
 int tlb_quiescence_check(u64_t unmap_time);
 
+/* Retype something back to frame */
 int
-retypetbl_retype2frame(void *pa)
+retypetbl_retype2frame(void *pa, u32_t order)
 {
-	struct retype_info_glb *glb_retype_info;
+	struct retype_entry_glb *glb_retype_info;
+	struct retype_entry     *retype_info;
+	struct retype_entry     temp, old_temp;
 	union refcnt_atom       local_u;
 	u32_t                   old_v, idx;
 	u64_t                   last_unmap;
-	int                     cpu, ret, ref_sum;
+	int                     i, j, cpu, ret, sum;
 	mem_type_t              old_type;
 
 	PA_BOUNDARY_CHECK();
 
 	idx = GET_MEM_IDX(pa);
 	assert(idx < N_MEM_SETS);
 
-	glb_retype_info = GET_GLB_RETYPE_ENTRY(idx);
-	old_type        = glb_retype_info->type;
-	/* only can retype untyped mem sets. */
+	/* Exact page granularity */
+	glb_retype_info = GET_GLB_RETYPE_ENTRY(idx, order);
+	old_type        = glb_retype_info->refcnt_atom.type;
+	/* Only can untype typed mem sets */
 	if (unlikely(old_type != RETYPETBL_USER && old_type != RETYPETBL_KERN)) return -EPERM;
 
-	/* lock down the memory set we are going to retype. */
-	ret = retypetbl_cas(&(glb_retype_info->type), old_type, RETYPETBL_RETYPING);
-	if (ret != CAS_SUCCESS) return ret;
+	/* lock down the memory set we are going to untype */
+	old_temp.refcnt_atom.v = glb_retype_info->refcnt_atom.v;
+	temp.refcnt_atom.v = old_temp.refcnt_atom.v;
+	temp.refcnt_atom.type = RETYPETBL_RETYPING;
 
-	last_unmap = 0;
-	for (ref_sum = 0, cpu = 0; cpu < NUM_CPU; cpu++) {
-		/* Keep in mind that, ref_cnt on each core could be
-		 * negative. */
-		old_v = local_u.v = retype_tbl[cpu].mem_set[idx].refcnt_atom.v;
-		assert(local_u.type == old_type || local_u.type == RETYPETBL_RETYPING);
+	ret = retypetbl_cas((u32_t*)glb_retype_info, old_temp.refcnt_atom.v, temp.refcnt_atom.v);
+	if (ret != CAS_SUCCESS) return ret;
 
-		local_u.type = RETYPETBL_RETYPING;
-		ret          = retypetbl_cas(&(retype_tbl[cpu].mem_set[idx].refcnt_atom.v), old_v, local_u.v);
-		if (ret != CAS_SUCCESS) cos_throw(restore_all, -ECASFAIL);
+	for (i = 0; i < NUM_CPU; i++) {
+		retype_info = GET_RETYPE_CPU_ENTRY(i, idx, order);
+		old_temp.refcnt_atom.v = retype_info->refcnt_atom.v;
+		temp.refcnt_atom.v = retype_info->refcnt_atom.v;
+		temp.refcnt_atom.type = RETYPETBL_RETYPING;
+		/* See if we can successfully lock it */
+		if (retypetbl_cas(&(retype_info->refcnt_atom.v), old_temp.refcnt_atom.v, temp.refcnt_atom.v) != CAS_SUCCESS) cos_throw(err, -EINVAL);
```

Comment:
Since we know that we are retyping, the only conflict here is with parallel maps, right?  In that case, can we confer this into an `faa` to convert types that we know will not conflict with other operations as the parallel faas (in the mapper which *do* require atomics) will be on different bits (i.e. for maps).